### ISSUE TITLE
[Feature Request]: Copy and Pasting in FTBQuests

### ISSUE THREAD

**Issue by SPYROHAWK** [FTBTeam/FTB-Mods-Issues#791]
### Mod

FTBQuests [Fabric]

### Description

I am using FTBQuests for Fabric 1.19.2 and I found there's a serious capability that is missing, and that is copy and paste. This comes in two forms: for quests, and within quests.

Let's start with what I mean by "within quests" because that's more intuitive. All text boxes within quest creation do not allow for copy and pasting. This is especially an issue for the command quest reward output. You make a quest that executes a command as a reward. When typing into the text box, it does not have the auto-fill feature like a normal command block, so trying to write the entire command from scratch is more difficult. So maybe you want to write it in a command block, then copy it and paste it in. Or maybe you are using a website for generating the command, and want to copy and paste it in. You can't. The text field for the command reward does not allow pasting. This is true for all text fields that I have found in the mod, with the exception of the quest descriptions. It is very frustrating and I am surprised this was not been fixed sooner.

The second variation of this issue, which I called "for quests", is less intuitive but a bigger issue for me personally. Simply, you can't copy and paste quests. Lets say I want to create 30 quests. All of these quests need to be repeatable and optional, the same dependencies, have an item input and a location input, and have an item output. The item input and output are changing, but everything else is the same. In order to do this, I need to go into the edit settings for all 30 quests and hit the optional and repeatable settings, add the dependencies, and then go into the quest and add the item input and output and locations, and fill out the 6 text boxes that define the location quest. That's a lot. It seems like there would be an easier way. But what if you could copy and paste? Well, I could create the one quest, select it by control-clicking it, control C to copy it, and control V to paste it 29 times. Now I have 30 identical quests, and I just need to go in and change the item input and outputs. So much easier.

### Describe the solution you'd like

For the first part, pretty much just allow copy and paste keybinds (control A + control C, control V [or command instead of control on a mac]) to work in the text boxes in quests.

Got the second part, the ability to copy and paste quests (either with control C and control V, or by an in-menu function of right-clicking a selected quest and having a copy option) to create an exact duplicate of the quest with the exact same settings, inputs, outputs, etc. This would be a huge timesaver for making a bunch of very similar quests with minor changes.

### Describe alternatives you've considered

I was thinking about using a reward table for quest grouping, but that still really doesn't solve the issue.

**Comment by desht** [comment]
>  All text boxes within quest creation do not allow for copy and pasting. 

Not true. All text boxes already allow pasting copied text via the standard Ctrl+V sequence. Ctrl+C, Ctrl+A work too.

> Simply, you can't copy and paste quests.

Also not true. You can copy any quest by right-clicking it, and choosing "Copy ID".  Then right click elsewhere and choose either "Paste Quest" or "Paste Quest Link", depending on whether you want a new quest or just a link to an existing quest.

**Comment by SPYROHAWK** [comment]
> > All text boxes within quest creation do not allow for copy and pasting.
> 
> Not true. All text boxes already allow pasting copied text via the standard Ctrl+V sequence. Ctrl+C, Ctrl+A work too.

Ah, perhaps this is a mac specific issue, because it uses command instead of control. I know you can't right click in the text boxes for the copy and pasting, and Command+V does not work.


> > Simply, you can't copy and paste quests.
> 
> Also not true. You can copy any quest by right-clicking it, and choosing "Copy ID". Then right click elsewhere and choose either "Paste Quest" or "Paste Quest Link", depending on whether you want a new quest or just a link to an existing quest.

Huh, I don't know how I missed that functionality. I'm going to need to take a look at that...

**Comment by SPYROHAWK** [comment]
> > All text boxes within quest creation do not allow for copy and pasting.
> 
> Not true. All text boxes already allow pasting copied text via the standard Ctrl+V sequence. Ctrl+C, Ctrl+A work too.
> 
> > Simply, you can't copy and paste quests.
> 
> Also not true. You can copy any quest by right-clicking it, and choosing "Copy ID". Then right click elsewhere and choose either "Paste Quest" or "Paste Quest Link", depending on whether you want a new quest or just a link to an existing quest.

Sorry, I may just be completely blind.

I find the Copy ID button:
<img width="1440" alt="Screenshot 2023-05-11 at 10 20 40 AM" src="https://github.com/FTBTeam/FTB-Mods-Issues/assets/16589796/f97243f5-0b77-42a8-a5d0-0b45dc0ccecc">

But then I right click somewhere else, it just shows the normal quest creation options, not any option to paste the quest from the ID:
<img width="1512" alt="Screenshot 2023-05-11 at 10 19 42 AM" src="https://github.com/FTBTeam/FTB-Mods-Issues/assets/16589796/d6969202-ef00-4230-8fd8-6f27c200d4d5">

**Comment by desht** [comment]
What version of the mod are you running?

**Comment by desht** [comment]
Regarding Mac Control vs Command, I'll take a look at the code and see if there's anything specific in there. Hard to test though when I don't have a Mac to run on...

As for docs, yeah they're lacking at the moment.  But we do have someone currently working on much better docs for all the FTB mods, including Quests. So this will hopefully be in a much better place soon.

**Comment by SPYROHAWK** [comment]
> What version of the mod are you running?

FTBQuests (Fabric) for Minecraft 1.19.2
Mod version 1902.4.15-build.228

(I don't know if that's the right format for the mod version or not, that's the name of the jar file on curseforge)

**Comment by desht** [comment]
OK, you're up to date, so there's probably a need for some Mac fixes here. I'll take a look.

**Comment by adamico** [comment]
+1 to add a shortcut ctrl+v and ctrl+shift+v to paste quest without and with dependencies. It's annoying having to right click and select the different paste options. Mmh, maybe a separate feature request to add more keyboard shortcuts could be more appropriate because I can think of other useful missing shortcuts (select quest, ctrl right click another quest to add dependency from and to selected quest, for example)

**Comment by desht** [comment]
Yes, a separate feature request with a bunch of useful hotkey would be welcome.

**Comment by Slimik222** [comment]
The "paste" button disappears for me if the Quests additions mod is installed

**Comment by Slimik222** [comment]
Found it! copy the ID, right-click on the place where you want to paste and hold Shift and turn the mouse wheel
![image](https://github.com/FTBTeam/FTB-Mods-Issues/assets/91560161/b2d49630-4237-4f8f-8d8f-af018acef92f)

**Comment by MichaelHillcox** [comment]
Looks to have been added :+1:
